WIP: Download nomad logs from webui

swsnr opened the pull request:
Add buttons to download logs from the webui:

![](https://user-images.githubusercontent.com/224922/29411634-fb1f049c-8355-11e7-94d3-e6cb50133903.PNG)

The download link gets the last 500 KiB of the log; the `offset` query parameter lets you go back further. 

Builds upon #277 which needs to be merged before this one.

swsnr committed:
Split Instances class out of instances services

Add a dedicated class for instances functionality that directly builds upon Nomad

swsnr committed:
GH-44: Get logs from the nomad client

swsnr committed:
Extract API routes into API router

"Seal" API routes by returning 404 on unknown API routes to provide proper 404 responses to invalid API requests
and prevent fallback to other routes

swsnr committed:
GH-44: Add endpoint to download logs

swsnr committed:
GH-44: Add buttons to download logs from webui

swsnr committed:
Properly parse HTTP address of node

Inherit protocol schema from configured Nomad URL.

Introduce scala-uri and build Nomad API urls with its nice DSL.

swsnr committed:
Limit the size of downloaded log files

swsnr committed:
Add query parameter for log offset

In webui get the last 500 KiB from the log.

swsnr committed:
Fix compilation errors in integration test

FRosner commented on server/src/it/scala/de/frosner/broccoli/nomad/NomadHttpClientIntegrationSpec.scala:
```diff
@@ -49,15 +51,15 @@ class NomadHttpClientIntegrationSpec
               response
             })
-          allocations <- client.getAllocationsForJob(identifier)
+          allocations <- client.getAllocationsForJob(shapeless.tag[Job.Id](identifier))
```
Thanks for introducing tagges types!

FRosner commented on server/src/main/scala/de/frosner/broccoli/nomad/models/Allocation.scala:
```diff
@@ -2,20 +2,31 @@ package de.frosner.broccoli.nomad.models
-                            taskStates: Map[String, TaskStateEvents])
+final case class Allocation(
+    id: String @@ Allocation.Id,
+    jobId: String @@ Job.Id,
```
Shall we create a type alias for that?

`type JobId = String @@ Job.Id`?

swsnr replied:
Personally I don't need them, I find they introduce a lot of clutter and boilerplate without any significant benefit.  These types are short; aliasing them won't save much will it?

FRosner commented on server/src/main/resources/routes:
```diff
@@ -3,7 +3,10 @@
+-> /api/v1  de.frosner.broccoli.routes.ApiV1Router
+
+# Downloads
+-> /downloads/ de.frosner.broccoli.routes.DownloadsRouter
```
What is the reasoning behind not making this part of our versioned API? IMHO downloading logs for an instance task allocation is not different from asking for its parameters. Also `/downloads` sounds like its a folder mapping where you can download static files. But it's actually something to refresh to get more logs.

@Gerrrr @lunaryorn what do you think?

Gerrrr replied:
Yep, I think that it should be at `/api/v1` since it's a part of the API.

swsnr replied:
I don't think that this is an API endpoint.  

Can I send it JSON? Does it return JSON?  Do we use it as an API endpoint?  I don't think so: We use it as a static endpoint for the webui, and it's not intended to serve structured data but rather human-readable "logs" whatever they might contain.

As for `/downloads/`: It needs to be in a separate URL prefix to allow proper 404s for non-existing logs, and I couldn't come up with a better prefix.  While I agree that it's not a "static file" it's not a "dynamic view" either: The log won't refresh, so in a way you're really "downloading" a snapshot of the log at this particular point of time.

FRosner replied:
It does not send JSON and it does not receive JSON. It receives a GET request and returns a text. But the format does not determine whether it is part of an API or not. For me there are two types of endpoints:

1. Endpoints that are meant to be accessed by a browser, serving a webpage with a layout, etc. This is the index.html basically. I'd call those part of the UI.
2. Endpoints that are meant to be accessed by programs, CLIs, scripts. This is useful for interacting with Broccoli without using a browser. I'd call those part of the API.

While we are certainly using the endpoint through a browser this is just temporary. We do it because we didn't include the logs directly in the UI. But what you get there is a resource and the protocol should be versioned. Why do I care? If someone uses Broccoli from a script or from the command line and we change the API (e.g. no longer return text but JSON), then it should be reflected in a new version of the API.

I agree that not sending the logs in JSON is a bit inconsistent. But we can make that call. We just need to document it.

I didn't fully get the separate URL prefix issue. Each endpoint can return 404 individually. If you request an instance that does not exist, you get 404. Same for logs.

swsnr replied:
> While we are certainly using the endpoint through a browser this is just temporary.

If it's just temporary what's going to replace it? What will replace this endpoint to provide the ability to, well, "download" a snapshot of the log to view (grep, less, etc.) it on the local system?

I'm aware that we'll stream logs at some point but is that a replacement for the ability to get the log on the local system?  I can't `grep` in the webui, can I?

> But what you get there is a resource and the protocol should be versioned.  Why do I care? If someone uses Broccoli from a script or from the command line and we change the API […]

The resource offers logs as they are on nomad node.  How could this endpoint change?  Or, in other words, if we wanted to change this endpoint wouldn't that rather be a change in the semantics and thus be a wholly new endpoint?

> I didn't fully get the separate URL prefix issue. Each endpoint can return 404 individually. If you request an instance that does not exist, you get 404. Same for logs.

Try `curl localhost:9000/this-does-not-exist` versus `curl localhost:9000/downloads/this-does-not-exist`:  The latter returns an empty 404, whereas the former returns a HTML page.

Since we already have a top-level catch-all route (for assets) that's only possible with a separate prefix.  

For the same reason there's a dedicated API router now.

FRosner replied:
The endpoint can stay as it is very useful. What I'm saying is that it is not part of the UI. And anything that is not part of the UI, is part of the API, in my opinion. Then it makes more sense to add it as part of the API from a routing point of view, no?

I don't know how this endpoint is going to evolve but I see advantages of having a stable API and versioning it at the same time. This does not conflict IMO.

I didn't get the last part with the curls. If the dedicated API router is there why doesn't it return 404 if the API route does not exist?

@Gerrrr what do you think?

swsnr replied:
> What I'm saying is that it is not part of the UI. 

The UI uses it as of now.

If we moved it to the stable and versioned API we would have to provide more guarantees than we do currently.  We would need to introduce content negotiation to support JSON and plain text, we would have to specify parameter ranges for the log size, etc.

All of this is missing currently, because I treated it as a UI endpoint.

As an example of the implication of "stabilizing" this endpoint:  What happens if the user tries to load a 2GB log file?  Currently we allocate a string for the entire logfile (because that's simple), but since the UI only gets the last some KBs there's no harm.  If we stabilize the endpoint we need to refactor the code to stream the log file from Nomad to the frontend to avoid excessive memory allocation for large log files.  That's easy for plain text, but I don't know whether Play can encode large strings in a streaming fashion. 

If it can't we suddenly need to two code paths for different content types or impose a maximum limit on the log size, but then how to we pick this limit?  How to we make sure the limit cuts the log file correctly, ie doesn't split in the middle of surrogate pairs rendering the log unreadable to the client?

If we can provide satisfying answers to these questions we can make it part of the stable API.  If we can't we shouldn't, because—since it is a "stable" API—any bug we make becomes part of the interface and can't be fixed without introducing new API versions or potentially breaking clients.

> I don't know how this endpoint is going to evolve.

Then why plan for change management?    Why not keep it as UI endpoint that lets us change it if *we* need it, without consideration for compatibility?

Don't be too fast with declaring something "stable" without giving it due time to ripe :wink:

> I didn't get the last part with the curls. If the dedicated API router is there why doesn't it return 404 if the API route does not exist?

It does, but I think we're talking past each other.  I referred to why I introduced `/downloads` as a prefix instead of adding a top-level route, not about `/downloads/` versus `/api/`.

As I said I do not think that this endpoint is part of the API.

FRosner replied:
Ok. To summarize we don't have an API endpoint to get logs at the moment, only a very ugly part of the UI. Correct?

Works for me! 👍

swsnr replied:
@FRosner Yes that's the situation from my point of view.  From this state we can slowly refine to  get a nicer UI… an API for logs is less important to me as long as we don't have specific need.

FRosner replied:
Agreed thank you!

FRosner commented on server/src/main/scala/de/frosner/broccoli/nomad/NomadModule.scala:
```diff
@@ -37,5 +38,5 @@ class NomadModule extends AbstractModule with ScalaModule {
   @Singleton
   def provideNomadClient(config: NomadConfiguration, wsClient: WSClient): NomadClient =
-    new NomadHttpClient(config.url, wsClient)(play.api.libs.concurrent.Execution.defaultContext)
+    new NomadHttpClient(Uri.parse(config.url), wsClient)(play.api.libs.concurrent.Execution.defaultContext)
```
Does this change resolve the problem reported in https://github.com/FRosner/cluster-broccoli/issues/81?

swsnr replied:
Perhaps… I don't know what message `Uri.parse` throws.

swsnr replied:
IOW, I didn't make this change to address #81, but just to build URLs in the client in a simpler way, and I don't think that it'll result in better error messages.

To address #81 you'd need to validate the URL upfront while parsing the configuration, to be able to throw an error message that can still refer to the path of the malformed value.

FRosner reviewed (changes requested):
Looks really nice! Thanks a lot. I assume that after rebasing I anyway will have to review again so let's discuss the few comments I had before merging. I also wanted to look at the UI part and maybe make a few small changes to the layout.

swsnr commented:
@FRosner What are these changes?  Why did you push them to this branch?

Wasn't too nice, to be honest… the pull request was still WIP, and not ready.  Now I've got to rebase your changes on top of my local changes 😒

swsnr commented on webui/src/Views/InstanceView.elm:
```diff
@@ -45,6 +45,14 @@ jobControlsColumnWidth =
+
+
+allocationLogColumnWidth =
+    110
```
Why do we need fixed widths for columns?

FRosner replied:
If a column has fixed content I prefer to have the width fixed. This way the layout is more consistent across multiple of those tables.

swsnr commented on webui/src/Views/InstanceView.elm:
```diff
@@ -412,7 +455,29 @@ instanceAllocationRow index ( taskName, allocation ) =
+                [ span [ class ("label " ++ labelKind) ] [ text description ] ]
+            , td
+                [ width allocationLogColumnWidth
+                , class "text-center"
```
Why center?

FRosner replied:
It looks better.

FRosner commented:
@lunaryorn the changes are related to the UI part of this change. I pushed them to this branch in order to support you. I understand that it is WIP but you requested my review so that's what I did. During the review I found something I didn't like so I changed it. If you struggle with rebasing you can also cherry-pick?

swsnr committed:
Inject controllers as providers

Avoid eager initialization and triggering side effects when creating
applications in the test suite.

We shouldn't have to do this, but unfortunately all services trigger
side effects as part of initialization

swsnr committed:
Change button colour to avoid conflicts with label colours

swsnr committed:
Center state and download columns

swsnr committed:
Use fixed layout for tasks table

Makes sure that all tables of this kind have the same column widths.

swsnr commented:
I've force-pushed to fix the test, use a different colours for the buttons (namely `info`) and use fixed table layout for the tasks table to make sure that all those tables have the same column widths:

<img width="1181" alt="bildschirmfoto 2017-08-18 um 14 33 39" src="https://user-images.githubusercontent.com/224922/29459094-723e8ae4-8422-11e7-9ef2-c8197bcb9c64.PNG">

swsnr commented:
@FRosner In contrast to fixed column width this approach equally distributes whitespace in the table.  Large whitespace between allocation and state makes it hard to scan the table, imho, as the state's visually disconnected from the allocation cell.

swsnr commented:
Smaller display:

<img width="671" alt="bildschirmfoto 2017-08-18 um 14 38 10" src="https://user-images.githubusercontent.com/224922/29459210-e9374c3a-8422-11e7-8949-b76d03ea69c7.PNG">

FRosner commented:
When you look at this table, I find it quite visually imbalanced. As discussed, let's reorder the table, putting allocation id and state in the first two columns, placing the task name in the middle claiming as much space as possible, and aligning the log buttons to the right.

![image](https://user-images.githubusercontent.com/3427394/29459456-e9e80f6a-8423-11e7-858a-d8569bd20954.png)

The button color should not be more offensive than the start and stop buttons, IMHO. I liked them white, like all the other buttons.

Also I think that "Logs" describes the column good enough, because technically these are not files. They are log streams and you get a snapshot when you click on it. When you refresh, you get a new snapshot. What bothers you with "Logs"?

swsnr committed:
Reorder task view columns

Move task name to the center and make it take all horizontal space

swsnr committed:
Use default colour for log buttons

swsnr committed:
Center log buttons and task states

swsnr commented:
How about this?

<img width="729" alt="bildschirmfoto 2017-08-21 um 10 14 30" src="https://user-images.githubusercontent.com/224922/29509778-a8ee75de-8659-11e7-9d0d-1dec14b98198.PNG">

swsnr commented:
@FRosner Ping :)

FRosner commented:
Sorry :D

I'm looking right now.

FRosner commented:
Awesome!
